### PR TITLE
Explain game tiles and entities in editor (fixes #1600)

### LINKED ISSUE #1600: Explain game tiles in client
instead of just in https://ddnet.tw/explain/?layer=game , should show an explanation when hovering other the respective game tile in editor.

### PR DESCRIPTION
**Pull request by def-**
- Explanations by Lady Saavik taken from https://ddnet.tw/explain/
- CCW/CW were mixed up in enum names, fixed
- Make sure that the texts fits, otherwise reduce font size
- Still need explanations for Portal tiles, @Ryozuki could you provide them?

### CODE REVIEW

File: src/game/editor/explanations.cpp [hunk 1]
```diff
@@ -0,0 +1,451 @@
...
+		if(Layer == LAYER_SWITCH)
+			return "FREEZE: Freezes tees for defined amount of seconds.";
+		break;
+	case TILE_UNFREEZE:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "UNFREEZE: Unfreezes tees immediately.";
+		break;
+	case TILE_TELEINEVIL:
+		if(Layer == LAYER_TELE)
+			return "EVIL TELEPORT: After falling into this tile, tees appear on TO with the same number. Speed and hooks are deleted.";
```
Comment (bojidar-bg) [inline src/game/editor/explanations.cpp]: Suggestion: "**RED** TELEPORT: After falling into this tile, tees appear on TO with the same number. Speed and hooks are **cleared**."

File: src/game/editor/explanations.cpp [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+		break;
+	case TILE_EHOOK_END:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "ENDLESS HOOK: Endless hook has been deactivated.";
+		break;
+	case TILE_HIT_START:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "HIT OTHERS: You can hit others.";
+		if(Layer == LAYER_SWITCH)
+			return "HIT OTHERS: You can activate hitting others for single weapons.";
```
Comment (bojidar-bg) [inline src/game/editor/explanations.cpp]: Suggestion: "HIT OTHERS: You can activate hitting others for single weapons, **using the delay number to specify which ones**" (and below)
If too long: "HIT OTHERS: You can activate hitting others for single weapons, **using delay to select which**"

File: src/game/editor/explanations.cpp [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+		break;
+	case TILE_SOLO_START:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "SOLO: You are now in a solo part.";
+		break;
+	case TILE_SOLO_END: // also TILE_SWITCHTIMEDOPEN
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "SOLO: You are now out of the solo part.";
+		if(Layer == LAYER_SWITCH)
+			return "TIME SWITCH: Activates switch (e.g. closes door) with the same number for set amount of seconds.";
```
Comment (bojidar-bg) [inline src/game/editor/explanations.cpp]: Suggestion: "TIME SWITCH: Activates switch (e.g. closes door) with the same number for **a** set amount of seconds." (and below)

File: src/game/editor/explanations.cpp [hunk 4]
```diff
@@ -0,0 +1,451 @@
...
+		if(Layer == LAYER_SWITCH)
+			return "TIME SWITCH: Deactivates switch (e.g. opens door) with the same number for set amount of seconds.";
+		break;
+	case TILE_SWITCHOPEN:
+		if(Layer == LAYER_SWITCH)
+			return "SWITCH: Activates switch (e.g. closes door) with the same number.";
+		break;
+	case TILE_SWITCHCLOSE:
+		if(Layer == LAYER_SWITCH)
+			return "SWITCH: Deactivates switch (e.g. closes door) with the same number.";
```
Comment (bojidar-bg) [inline src/game/editor/explanations.cpp]: Suggestion: "SWITCH: Deactivates switch (e.g. **opens** door) with the same number." (similar to above)

File: src/game/editor/explanations.cpp [hunk 5]
```diff
@@ -0,0 +1,451 @@
...
+		break;
+	case TILE_TELEIN:
+		if(Layer == LAYER_TELE)
+			return "TELEPORT: After falling into this tile, tees appear on TO with the same number. Speed and hook are kept.";
```
Comment (bojidar-bg) [inline src/game/editor/explanations.cpp]: Suggestion: "**BLUE** TELEPORT: After falling into this tile, tees appear on TO with the same number. Speed and hook are kept."
Alternatively, "**GOOD** TELEPORT"

File: src/game/editor/explanations.cpp [hunk 6]
```diff
@@ -0,0 +1,451 @@
...
+		break;
+	case TILE_TELEOUT:
+		if(Layer == LAYER_TELE)
+			return "TELEPORT: Teleport destination tile for FROMs, WEAPON & HOOK TELEPORTs with the same numbers.";
```
Comment (bojidar-bg) [inline src/game/editor/explanations.cpp]: Suggestion: "TELEPORT **TO**: Teleport destination tile for FROMs, WEAPON & HOOK TELEPORTs with the same numbers."

File: src/game/editor/explanations.cpp [hunk 7]
```diff
@@ -0,0 +1,451 @@
...
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "STOPPER: You can hook and shoot through it. You can't go through it against the arrows.";
+		break;
+	case TILE_STOPA:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "STOPPER: You can hook and shoot through it. You can't go through it.";
+		break;
+	case TILE_TELECHECKINEVIL:
+		if(Layer == LAYER_TELE)
+			return "CHECKPOINT EVIL TELE: Send tees to CTO with the same number as the last touched TELEPORT CHECKPOINT. Speed and hook are deleted.";
```
Comment (bojidar-bg) [inline src/game/editor/explanations.cpp]: (as above) Suggestion: "**RED** CHECKPOINT TELE**PORT**…are **cleared**."

File: src/game/editor/explanations.cpp [hunk 8]
```diff
@@ -0,0 +1,451 @@
...
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "COLLISION: You can't collide with others.";
+		break;
+	case TILE_SUPER_END:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "SUPER JUMP: You don't have unlimited air jumps.";
+		break;
+	case TILE_JETPACK_END:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "JETPACK: You lost your jetpack gun.";
```
Comment (bojidar-bg) [inline src/game/editor/explanations.cpp]: Suggestion: "JETPACK **OFF**: You **lose** your jetpack gun."

File: src/game/editor/explanations.cpp [hunk 9]
```diff
@@ -0,0 +1,451 @@
...
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "SPAWN: Not used in DDRace. Here spawn blue team members.";
+		break;
+	case ENTITY_OFFSET + ENTITY_FLAGSTAND_RED:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "FLAG: Not used in DDRace. Place where red team flag is.";
+		break;
+	case ENTITY_OFFSET + ENTITY_FLAGSTAND_BLUE:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "FLAG: Not used in DDRace. Place where red team blue is.";
```
Comment (bojidar-bg) [inline src/game/editor/explanations.cpp]: Suggestion: "FLAG: Not used in DDRace. Place where **blue** team **flag** is.";

File: src/game/editor/explanations.cpp [hunk 10]
```diff
@@ -0,0 +1,451 @@
...
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT || Layer == LAYER_SWITCH)
+			return "GRENADE LAUNCHER: Throws exploding bullets. Also known as rocket.";
+		break;
+	case ENTITY_OFFSET + ENTITY_POWERUP_NINJA:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT || Layer == LAYER_SWITCH)
+			return "NINJA: Makes you invisible in the darkest nights.";
+		break;
+	case ENTITY_OFFSET + ENTITY_WEAPON_RIFLE:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT || Layer == LAYER_SWITCH)
+			return "RIFLE: Unfreezes the tee. Bounces off the walls. Also known as laser.";
```
Comment (bojidar-bg) [inline src/game/editor/explanations.cpp]: Suggestion: "RIFLE: Unfreezes tee**s hit**. Bounces off the walls. Also known as laser.""

File: src/game/editor/explanations.cpp [hunk 11]
```diff
@@ -0,0 +1,451 @@
...
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "SPEEDER: Causes weapons, SHIELD, HEART and SPINNING LASER to move quickly.";
+		break;
+	case TILE_TUNE1:
+		if(Layer == LAYER_TUNE)
+			return "TUNE ZONE: Area where defined tunes work.";
+		break;
+	case TILE_OLDLASER:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "OLD LASER: Shotgun drags others always towards the shooter, even after having bounced. Shooter can't hit themselves. Place only one tile somewhere on the map.";
```
Comment (bojidar-bg) [inline src/game/editor/explanations.cpp]: Suggestion: "OLD **SHOTGUN**: "

File: src/game/editor/explanations.cpp [hunk 12]
```diff
@@ -0,0 +1,451 @@
...
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT || Layer == LAYER_SWITCH)
+			return "SPINNING LASER: Tile where freezing laser (made with LASER LENGTH) begins. Clockwise, medium speed.";
+		break;
+	case ENTITY_OFFSET + ENTITY_LASER_FAST_CW:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT || Layer == LAYER_SWITCH)
+			return "SPINNING LASER: Tile where freezing laser (made with LASER LENGTH) begins. Clockwise, fast.";
+		break;
+	case ENTITY_OFFSET + ENTITY_LASER_SHORT:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT || Layer == LAYER_SWITCH)
+			return "LASER LENGTH: Combined with DOOR or SPINNING LASER, makes it 3 tiles long.";
```
Comment (bojidar-bg) [inline src/game/editor/explanations.cpp]: Suggestion: "LASER LENGTH: **Put next to** DOOR or SPINNING LASER…" (and below)

File: src/game/editor/explanations.cpp [hunk 13]
```diff
@@ -0,0 +1,451 @@
...
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT || Layer == LAYER_SWITCH)
+			return "LASER LENGTH: Combined with DOOR or SPINNING LASER, makes it 6 tiles long.";
+		break;
+	case ENTITY_OFFSET + ENTITY_LASER_LONG:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT || Layer == LAYER_SWITCH)
+			return "LASER LENGTH: Combined with DOOR or SPINNING LASER, makes it 9 tiles long.";
+		break;
+	case ENTITY_OFFSET + ENTITY_LASER_C_SLOW:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT || Layer == LAYER_SWITCH)
+			return "LASER LENGTH CHANGE: Combined with LASER LENGTH, causes it to length and shorten constantly. Works only on (NON-)SPINNING LASER, not on DOOR. Lengthen, slow.";
```
Comment (bojidar-bg) [inline src/game/editor/explanations.cpp]: As above, suggestion: "LASER LENGTH CHANGE:…**Put next to**…"

File: src/game/editor/explanations.cpp [hunk 14]
```diff
@@ -0,0 +1,451 @@
...
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT || Layer == LAYER_SWITCH)
+			return "DOOR: Combined with LASER LENGTH creates doors. Doesn't allow to go through it (only with NINJA).";
+		break;
+	case TILE_TELE_GUN_ENABLE:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "TELEGUN: Turn gun on as telegun weapon.";
+		break;
+	case TILE_TELE_GUN_DISABLE:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "TELEGUN: Turn gun off as telegun weapon.";
```
Comment (bojidar-bg) [inline src/game/editor/explanations.cpp]: Suggestion: "TELEGUN **OFF**:…" (and below, 3 total)

File: src/game/editor/explanations.cpp [hunk 15]
```diff
@@ -0,0 +1,451 @@
...
+		break;
+	case TILE_ALLOW_BLUE_TELE_GUN:
+		if(Layer == LAYER_FRONT)
+			return "TELEGUN: Place on top of a collision tile, activates a spot to teleport to, preserves movement.";
+		if(Layer == LAYER_SWITCH)
+			return "TELEGUN: Place on top of a collision tile, activates a spot to teleport to, preserves movement, for single weapons.";
+		break;
+	case TILE_TELE_GRENADE_ENABLE:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "TELEGUN: Turn grenade on as telegun weapon.";
```
Comment (bojidar-bg) [inline src/game/editor/explanations.cpp]: Suggestion: "TELE**GRENADE**:…" (and below)

File: src/game/editor/explanations.cpp [hunk 16]
```diff
@@ -0,0 +1,451 @@
...
+		break;
+	case TILE_TELE_GRENADE_DISABLE:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "TELEGUN: Turn grenade off as telegun weapon.";
+		break;
+	case TILE_TELE_LASER_ENABLE:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "TELEGUN: Turn laser on as telegun weapon.";
```
Comment (bojidar-bg) [inline src/game/editor/explanations.cpp]: Suggestion: "TELE**LASER**:…" or "TELE**RIFLE**:…" (and below)

File: src/game/editor/explanations.cpp [hunk 17]
```diff
@@ -0,0 +1,451 @@
...
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "TELEGUN: Turn gun on as telegun weapon.";
+		break;
+	case TILE_TELE_GUN_DISABLE:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "TELEGUN: Turn gun off as telegun weapon.";
+		break;
+	case TILE_ALLOW_TELE_GUN:
+		if(Layer == LAYER_FRONT)
+			return "TELEGUN: Place on top of a collision tile, activates a spot to teleport to, cancels movement.";
```
Comment (bojidar-bg) [inline src/game/editor/explanations.cpp]: Suggestion: "**ALLOW** TELEGUN: …" (and below)

File: src/game/editor/explanations.cpp [hunk 18]
```diff
@@ -0,0 +1,451 @@
...
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "SPEEDER: Causes weapons, SHIELD, HEART and SPINNING LASER to move quickly.";
+		break;
+	case TILE_PENALTY:
+		if(Layer == LAYER_SWITCH)
+			return "PENALTY: Adds time to your current race time. Opposite of BONUS.";
+		break;
+	case TILE_NPC_END:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "COLLISION: You can't collide with others.";
```
Comment (bojidar-bg) [inline src/game/editor/explanations.cpp]: Suggestion: "COLLISION **OFF**: …" (and below, 4 times total)

File: src/game/editor/explanations.cpp (hunk 11 above)
Comment (bojidar-bg) [reply src/game/editor/explanations.cpp]: Also, suggestion: "**GLOBAL** OLD…" (and below, 6 times total)

File: src/game/editor/explanations.cpp [hunk 19]
```diff
@@ -0,0 +1,451 @@
...
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "HIT OTHERS OFF: Nobody can hit others. Place only one tile somewhere on the map.";
+		break;
+	case TILE_NPH:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "HOOK OTHERS OFF: Nobody can hook others. Place only one tile somewhere on the map.";
+		break;
+	case TILE_UNLOCK_TEAM:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "SPEEDER: Causes weapons, SHIELD, HEART and SPINNING LASER to move quickly.";
```
Comment (bojidar-bg) [inline src/game/editor/explanations.cpp]: Suggestion: "**UNLOCK TEAM: Forces the team to be unlocked.**"

File: src/game/editor/explanations.cpp [hunk 20]
```diff
@@ -0,0 +1,451 @@
...
+		if(Layer == LAYER_SPEEDUP)
+			return "SPEEDUP: Gives tee defined speed. Arrow shows direction and angle.";
+		break;
+	case TILE_TELECHECK:
+		if(Layer == LAYER_TELE)
+			return "CHECKPOINT TELEPORT: After having touched this tile, any CFRM will teleport you to CTO with the same number.";
+		break;
+	case TILE_TELECHECKOUT:
+		if(Layer == LAYER_TELE)
+			return "CHECKPOINT TELEPORT: Here tees will appear after touching TELEPORT CHECKPOINT with the same number and falling into CFROM TELEPORT.";
```
Comment (bojidar-bg) [inline src/game/editor/explanations.cpp]: Suggestion: "CHECKPOINT TELEPORT **TO**:…"

File: src/game/editor/explanations.cpp [hunk 21]
```diff
@@ -0,0 +1,451 @@
...
+		break;
+	case TILE_TELECHECKIN:
+		if(Layer == LAYER_TELE)
+			return "CHECKPOINT TELEPORT: Sends tees to CTO with the same number as the last touched TELEPORT CHECKPOINT. Speed and hook are kept.";
```
Comment (bojidar-bg) [inline src/game/editor/explanations.cpp]: Suggestion (as above): "**BLUE** CHECKPOINT TELEPORT:…", alternatively "**GOOD**"

File: src/game/editor/explanations.cpp [hunk 22]
```diff
@@ -0,0 +1,451 @@
...
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT || Layer == LAYER_SWITCH)
+			return "UNDEEP: Removes DEEP FREEZE effect.";
+		break;
+	case TILE_TELEINWEAPON:
+		if(Layer == LAYER_TELE)
+			return "WEAPON TELEPORT: Teleports bullets shot into it to TELEPORT TO, where it comes out. Direction, angle and length are kept.";
+		break;
+	case TILE_TELEINHOOK:
+		if(Layer == LAYER_TELE)
+			return "HOOK TELEPORT: Teleports hooks entered into it to TELEPORT TO, where it comes out. Direction, angle and length are kept.";
```
Comment (bojidar-bg) [inline src/game/editor/explanations.cpp]: Suggestion: "HOOK TELEPORT: Teleports hooks **entering** it to TELEPORT TO, where it comes out.
Direction, angle **,** (comma!) and length are kept."

File: src/game/editor/explanations.cpp [hunk 23]
```diff
@@ -0,0 +1,451 @@
...
+		break;
+	case TILE_WALLJUMP:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "WALLJUMP: Placed next to a wall. Enables to climb up the wall.";
```
Comment (bojidar-bg) [inline src/game/editor/explanations.cpp]: Suggestion: "…Enables **climbing** up the wall"

File: src/game/editor/explanations.cpp [hunk 24]
```diff
@@ -0,0 +1,451 @@
...
+		break;
+	case TILE_EHOOK_START:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "ENDLESS HOOK: Endless hook has been activated.";
+		break;
+	case TILE_EHOOK_END:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "ENDLESS HOOK: Endless hook has been deactivated.";
```
Comment (bojidar-bg) [inline src/game/editor/explanations.cpp]: Suggestion: "ENDLESS HOOK **OFF**: **Deactivates** endless hook." (similarly for the _START tile)

File: src/game/editor/explanations.cpp [hunk 25]
```diff
@@ -0,0 +1,451 @@
...
+	case TILE_ENTITIES_OFF_2:
+		return "ENTITIES OFF SIGN: Informs people playing with entities about important marks, tips, information or text on the map.";
+	// Entities
+	case ENTITY_OFFSET + ENTITY_SPAWN:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "SPAWN: Here tees will appear after joining the game or dying somewhere on the map.";
+		break;
+	case ENTITY_OFFSET + ENTITY_SPAWN_RED:
+		if(Layer == LAYER_GAME || Layer == LAYER_FRONT)
+			return "SPAWN: Not used in DDRace. Here spawn red team members.";
```
Comment (12pm) [inline src/game/editor/explanations.cpp]: It's used in ddrace, works like the normal spawn (same for blue spawn)

**Comment by def-** [comment]
What do they do?

**Comment by 12pm** [comment]
TILE_THROUGH_CUT: Shortcut for new hookthrough
TILE_THROUGH_ALL: Combined with a collision tile is new hookthrough, otherwise stops hooks
TILE_THROUGH_DIR: Same as TILE_THROUGH_ALL for one direction

**Comment by def-** [comment]
Done.

**Pull request merged**